### ISSUE TITLE
POC: State on Followers

### ISSUE THREAD

**Issue by Zelldon** [camunda-cloud/zeebe#7328]
**Description**

I want to create an POC based on the ZEP draft https://github.com/zeebe-io/enhancements/pull/16 in order to understand better what impact the changes will have and how it would look like. This should help us to plan this topic.

I will use this issue to collect results and observations.

**Comment by Zelldon** [comment]
### Day one 21.06.2021

Rewrote ReprocessingStateMachine, now replay only. Was replaced with ReplayStateMachine, property based replay tests running for ~ 1 hour successfully. Code https://github.com/camunda-cloud/zeebe/blob/zell-7328-poc-state-on-followers/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java

Error events have been ignored for now.

#### Findings

 * It seems that currently the blacklist is not recreated correctly.  
   * We need the possibility to apply error events, instead of retriggering the exception as we did before on reprocessing. 
   * For that we need a new method on the Blacklisting State to add PI via key, after that we can just read the PI key from the event and add it to the blacklist.
   * Might be something we need to do quite earlier, such that our tests are green? (But currently there is also nothing failing :thinking: )

#### Next Steps

  * Add commit listener to ReplayStateMachine, such that continuously replay is supported
  * Install ReplayStateMachine on followers, either have a conditional stmt in the StreamProcessor or a separate class to run/install it.
  * Reorganize ZeebePartition bootstrap, most of the service should be long living, either StreamProcessor is notified about role change or we close and open it on role change.
  * Dispatcher need to be reseted from outside. Might make sense for simplicity to just reopen log wrapper?
  * Install exporter service, which sends exporter positions and receives them to store them in state.

**Comment by Zelldon** [comment]
### Day two 22.06.2021

Today I made good progress. I implemented the rebuilding of the blacklist via replay error events. The continuously replay with commit listeners. I was able to rewrite (in a hacky way) the Partition bootstrapping, such some steps installed once as base and others if a role change happens. The StreamProcessor does continuously replay if the appropriated flag is set, otherwise it works as before. Replay can react on pausing and resuming, which means now replay can also be interrupted! I ignored Exporters for now (divide and conquer :crossed_swords: ) In order to reset the dispatcher, we reinstall the logstream abstraction for now.

Restart and reprocessing tests work in engine, broker and QA (for single node).

#### Findings
processingStateMachine.getLastSuccessfulProcessedEventPosition()

 * Currently only the LeaderRole implements ZeebeLogAppender. I had to change the AtomixLogStorage to use a supplier for the Logappender, otherwise I wouldn't be able to install it once. Alternative is to create an ReadOnlyLogStorage for the Follower, which would mean we always recreate it on transitioning.
 * New journal supports seekToLast - but this is not used by LogStream
 * I think the partition transition concept can be improved. The steps are highly coupled (have dependencies on each other). Maybe we have on base and then Leader/FolloerPartition which will be installed on transition.
 * On cont. replay (follower) what is the last written position? Which is consumed by async snapshot director. I would say from top of my head - the last follow up event position, which we have applied recently.

#### Next Steps
 
  * Need to get cluster QA tests running
  * Deletion of follower data currently not work, snapshots are not yet working correctly. 
  * Exporters

**Comment by npepinpe** [comment]
Great stuff!

Regarding the log appender - I would like to actually explore switching from a push model (i.e. push into the ZeebeLogAppender/LeaderRole new entries, which get appended to the executor queue) to a pull model, where the leader periodically consumes pending writes. This would not only simplify this (i.e. you don't need to know the log appender beforehand), but would also help stabilize the heartbeats even if we have tons of read at once by giving more control to the leader role when new writes are consumed.

Regarding the transitions - I talked briefly about it with Peter, and it's possible we don't really need transitions in the future. Right now, the follower ZeebePartition is just really reacting to new snapshots - it receives them, stores them, and triggers compaction. It doesn't really do anything else. Since this will be removed in the future, we can think of it more as components which are always there - the storage layer (snapshots, log), the Raft layer (for consensus), the ZeebeDb and Engine layer...they will be present for all. However, they should react to certain signals during transitions. So it might be worth revisiting the approach, and not have transitions which create/start/stop services, but rather using actors and message passing for control. I don't know if I'm explaining this properly, but we can talk about it :sweat_smile: (regarding the layers, I made a quick sketch to discuss with Peter about the bootstrapping so maybe it's clearer when I talk about layers if you see it, I added it below - keep in mind this is just an idea)

![Standalone Broker(3)](https://user-images.githubusercontent.com/43373/122951844-e158d480-d37d-11eb-8ca1-c8aef01522e2.png)

**Comment by Zelldon** [comment]
Hey @npepinpe thanks for your input!

>  where the leader periodically consumes pending writes.

If I get it right then it sounds like you want to remove one or two layers, since on the dispatcher we already have that case where someone has to consume the pending writes :) 

![Partition - Leader](https://user-images.githubusercontent.com/2758593/123048543-cc705580-d3fe-11eb-9ef7-51de4c9d82d6.png)

So the Leader would be directly in charge of consuming it. For me it would also be interesting to explore in the other direction, that the Leader and Follower notify  registered consumers about new appended entries, instead of using readers. In general I'm all in for simplifying and refactoring these things, but I think this is out of scope for this POC. 

> However, they should react to certain signals during transitions. So it might be worth revisiting the approach, and not have transitions which create/start/stop services

Yeah would be interesting to look at it and to verify how we can react then on role changes and related behavior changes (replay instead of processing, recreating of DB etc.)

**Comment by npepinpe** [comment]
Yes out of scope here for sure, I'm just mentioning it for the implementation, we could consider it if it simplifies things

**Comment by Zelldon** [comment]
### Day Three 23.06.2021

Today was a short day, but still I made some progress.

Implemented commit listeners for followers, followers didn't get commit positions updated on Logstream before. This caused issues on the ReplayStateMachine and SnapshotDirector. Snapshots are taken **now** on Followers and Compaction also works (no exporter details yet). QA Test ClusterDeletionTest works now. Seems we have some problems now on taking snapshots on single node, need to investigate.


Example output:
```
13:27:31.551 [Broker-1-SnapshotStore-1] [Broker-1-zb-fs-workers-0] DEBUG io.camunda.zeebe.logstreams.snapshot - Taking temporary snapshot into /tmp/junit17273420128997090342/1/data/raft-partition/partitions/1/pending/291-1-291-9223372036854775807.
13:27:31.552 [Broker-0-SnapshotStore-1] [Broker-0-zb-fs-workers-1] DEBUG io.camunda.zeebe.logstreams.snapshot - Taking temporary snapshot into /tmp/junit17273420128997090342/0/data/raft-partition/partitions/1/pending/289-1-289-9223372036854775807.
13:27:31.558 [Broker-2-SnapshotStore-1] [Broker-2-zb-fs-workers-0] DEBUG io.camunda.zeebe.logstreams.snapshot - Taking temporary snapshot into /tmp/junit17273420128997090342/2/data/raft-partition/partitions/1/pending/289-1-289-9223372036854775807.
13:27:31.561 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] INFO  io.camunda.zeebe.processor - #replayNextEvent
13:27:31.562 [Broker-2-StreamProcessor-1] [Broker-2-zb-actors-0] INFO  io.camunda.zeebe.processor - #replayNextEvent
13:27:31.582 [Broker-1-SnapshotDirector-1] [Broker-1-zb-fs-workers-0] INFO  io.camunda.zeebe.logstreams.snapshot - Finished taking temporary snapshot, need to wait until last written event position 292 is committed, current commit position is 293. After that snapshot will be committed.
13:27:31.588 [Broker-2-StreamProcessor-1] [Broker-2-zb-actors-0] INFO  io.camunda.zeebe.processor - #replayNextEvent
13:27:31.588 [Broker-2-StreamProcessor-1] [Broker-2-zb-actors-0] INFO  io.camunda.zeebe.processor - Has no next, retry in PT1S
13:27:31.591 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] INFO  io.camunda.zeebe.processor - #replayNextEvent
13:27:31.591 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] INFO  io.camunda.zeebe.processor - Has no next, retry in PT1S
13:27:31.592 [Broker-0-SnapshotDirector-1] [Broker-0-zb-fs-workers-1] INFO  io.camunda.zeebe.logstreams.snapshot - Finished taking temporary snapshot, need to wait until last written event position 292 is committed, current commit position is 292. After that snapshot will be committed.
13:27:31.620 [Broker-2-SnapshotDirector-1] [Broker-2-zb-fs-workers-0] INFO  io.camunda.zeebe.logstreams.snapshot - Finished taking temporary snapshot, need to wait until last written event position 292 is committed, current commit position is 292. After that snapshot will be committed.
13:27:31.630 [Broker-2-SnapshotDirector-1] [Broker-2-zb-fs-workers-0] DEBUG io.camunda.zeebe.logstreams.snapshot - Current commit position 293 >= 292, committing snapshot FileBasedTransientSnapshot{directory=/tmp/junit17273420128997090342/2/data/raft-partition/partitions/1/pending/289-1-289-9223372036854775807, checksum=1875841979, metadata=FileBasedSnapshotMetadata{index=289, term=1, processedPosition=289, exporterPosition=9223372036854775807}}.
13:27:31.630 [Broker-1-SnapshotDirector-1] [Broker-1-zb-fs-workers-0] DEBUG io.camunda.zeebe.logstreams.snapshot - Current commit position 293 >= 292, committing snapshot FileBasedTransientSnapshot{directory=/tmp/junit17273420128997090342/1/data/raft-partition/partitions/1/pending/291-1-291-9223372036854775807, checksum=3284416406, metadata=FileBasedSnapshotMetadata{index=291, term=1, processedPosition=291, exporterPosition=9223372036854775807}}.
13:27:31.634 [Broker-0-SnapshotDirector-1] [Broker-0-zb-fs-workers-1] DEBUG io.camunda.zeebe.logstreams.snapshot - Current commit position 293 >= 292, committing snapshot FileBasedTransientSnapshot{directory=/tmp/junit17273420128997090342/0/data/raft-partition/partitions/1/pending/289-1-289-9223372036854775807, checksum=303286653, metadata=FileBasedSnapshotMetadata{index=289, term=1, processedPosition=289, exporterPosition=9223372036854775807}}.
13:27:31.637 [Broker-1-SnapshotStore-1] [Broker-1-zb-fs-workers-0] INFO  io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore - Committed new snapshot 291-1-291-9223372036854775807
13:27:31.637 [Broker-0-SnapshotStore-1] [Broker-0-zb-fs-workers-1] INFO  io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore - Committed new snapshot 289-1-289-9223372036854775807
13:27:31.637 [Broker-2-SnapshotStore-1] [Broker-2-zb-fs-workers-1] INFO  io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore - Committed new snapshot 289-1-289-9223372036854775807
13:27:31.643 [Broker-1-DeletionService-1] [Broker-1-zb-actors-1] DEBUG io.camunda.zeebe.broker.logstreams.delete - Scheduling log compaction up to index 291
13:27:31.643 [Broker-2-DeletionService-1] [Broker-2-zb-actors-0] DEBUG io.camunda.zeebe.broker.logstreams.delete - Scheduling log compaction up to index 289
13:27:31.644 [] [raft-server-1-raft-partition-partition-1] DEBUG io.camunda.zeebe.journal.file.SegmentedJournal - raft-partition-partition-1 - Deleting log up from 1 up to 234 (removing 9 segments)
13:27:31.644 [] [raft-server-2-raft-partition-partition-1] DEBUG io.camunda.zeebe.journal.file.SegmentedJournal - raft-partition-partition-1 - Deleting log up from 1 up to 234 (removing 9 segments)
13:27:31.646 [Broker-0-DeletionService-1] [Broker-0-zb-actors-1] DEBUG io.camunda.zeebe.broker.logstreams.delete - Scheduling log compaction up to index 289
```
We can see on all nodes snapshot has been taken and log has been compacted. State has been built on all nodes (yes on followers as well). SnapshotReplication was already completely turned off.

### Findings

 * On recovering ZeebeDB on follower transition, we need to make sure (before copying the snapshot to runtime) to close the DB, otherwise we get weird behavior on taking snapshots etc. (if the old DB is referenced)
 * On develop the commit positions of the Logstream is only updated on leaders, if they append a new entry. This means the LogStream commitListeners are only called on Leaders. Solution is to use the commit listeners on the RaftContext, which are called if the real RAFT commit index is updated. In order to preserve the abstractions, we could add a method, which accepts adding new commit Listeners, on the [LogStorage interface](https://github.com/camunda-cloud/zeebe/blob/zell-7328-poc-state-on-followers/logstreams/src/main/java/io/camunda/zeebe/logstreams/storage/LogStorage.java#L50) (internally this delegates to the [RaftServer](https://github.com/camunda-cloud/zeebe/blob/zell-7328-poc-state-on-followers/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/AtomixLogStoragePartitionStep.java#L31-L37)). In the LogStreamImpl constructor we register a listener, like `logStorage.addCommitListener(this::setCommitPosition);`

### Next Steps

 * Exporter service to communicate exporter positions and receive exporter positions
 * Verify whether compaction on leader is a problem
 * Find a way to react on install requests on followers to recreated db and replay statemachine (this is probably one of the more problematic parts) 
 * Get QA Tests running
 * If QA Tests run - start with first benchmarks to see how the current state is

**Comment by Zelldon** [comment]
In the morning I just thought that I just want to try running a benchmark with the latest changes. Looks quite promising, take it with a grane of salt, since exporters changes are not there and it seems compaction doesn't work as expected, but still interesting to see.

In general it looks like we have a performance degradation of ~15 % (which is from my side expected). I thought of 10 - 20 %. The partitions are marked as unhealthy but they never become healthy might be related to https://github.com/camunda-cloud/zeebe/issues/6393 (and other related issues)

![24-06-general](https://user-images.githubusercontent.com/2758593/123223186-fa25cf00-d4d0-11eb-97d0-4aed3bde1d2c.png)

We can see that all nodes use the CPU's now heavily, maybe we can trim here a bit.

![24-06-res](https://user-images.githubusercontent.com/2758593/123223423-2c373100-d4d1-11eb-8a6e-09e73522b838.png)

In the RocksDB section we can see the metrics for all RocksDB instances

![24-06-rocks](https://user-images.githubusercontent.com/2758593/123223541-4244f180-d4d1-11eb-8a5c-6e53ddf35f88.png)

**Comment by npepinpe** [comment]
Memory usage seems fine as far as I can tell - though of course it means you need to provision a bit more memory than before since RocksDB needs memory to be efficient. I wonder what the I/O looks like? I'm also surprised that we're CPU bound, I wouldn't expect replay to have a big impact CPU wise.

**Comment by Zelldon** [comment]
### Day Four 24.06.2021

Today, was again an insightful day.

Most of the QA Tests are now running. I was able to deploy an more stable benchmark, which already gives some insights. I found a simple solution for reacting on install requests, whether the solution is appropriate needs to be discussed (properly via ZEP). 

### Findings

 * The **current transactions** and how we have implemented it, doesn't work well for the things we want to do
   * We have several smaller steps which are our dependencies/components we want to install, they are tightly coupled and depend on each other. Example: I can't install the `ZeebeDbPartitionStep` without previous installing `StateControllerPartitionStep` or I can't install `LogStreamPartitionStep` without `AtomixLogStoragePartitionStep` and I for the `StreamProcessorPartitionStep` I need a lot of other steps. This is totally in transparent and confusing (if you don't know how the components work, which I luckily do). 
   * If a transition is executed the previous steps are closed. This caused in my implementation to weird behavior. We wanted to reset the Dispatcher for the Leaders, only. This was the reason we installed the `LogStreamPartitionStep` again, such that the `LogStream` is reset. If the transition from Leader to Follower happens the logstream will be closed. This means we need on the Follower also such an step to recreate the Logstream. 
   * Ideally we have a specific transitions which contain all of it inside, so it is clear which service/component is removed/recreated or installed. Like BoostrapLeader, BootstrapFollower, LeaderToFollower and FollowerToFollower transition objects 
   * In my opinion it makes totally sense to wait on the bootstrap rewrite to start with implementing the state on followers, since this will probably change also a lot and might cause a lot of conflicts. Furthermore it might sense to synchronize here more, so we already plan a bit what we need for state on followers. \cc @pihme @npepinpe 
 * I had a bug in my **commit listener implementation**. I set the commit position in the LogStream to the commit index, which is obviously not what we want. I changed the interface that we get the indexed entry, such that we can extract the highest position out of the committed entry. For Leader this is easy to call the listeners with that entry, on followers not. 
   * On followers we get an AppendEntry, which contain entries and the next commitIndex. The entries are most of the time unrelated to the commit index, since they already way ahead. We need on the followers a way to get, via the given new commit index, the indexed entry in order to call the listener and translate it to our highest entry position.  
   * In my POC I have a separate reader, which I move forward and read the next committed entry and give it the listener. I think this should be fine, since it is probably still in the page cache. **Alternative** we could build again an index/map which maps the indexes to either entries or directly to the highest positions.
 * The old **QA snapshot tests** need to be changed, since the snapshots are no longer equal (expected)
   * There is also a timing issue with taking snapshot on all nodes. It might happen that the followers didn't started with replay yet and the test already increased the snapshot interval time, so watch out.
 * **Receive InstallRequest on followers:**
   * As I mentioned earlier I thought this might be a problematic part. 
   * First I realized that we need to take care of the different snapshot listeners. We have one to compact the log, but also on the PassiveRole one to reset the complete log if the indexes are not close enough. We should verify whether this is an issue.
   * I started think about reacting on the snapshot and realized if we already reset the log on an install request and based on the ZEP need to restore our state (copy snapshot to runtime) and restart replay. The simplest solution would be to restart an Follower transition, which means we could in RAFT transition again to follower and install everything new. I implemented that approach in the POC, was just adjusting some checks. If you think about the roles as states in a state machine then it makes sense to take again a transition to restore the state.
![statemachine](https://user-images.githubusercontent.com/2758593/123279096-d3cf5600-d507-11eb-8795-dae59ad7f9f9.png)
 * StateController should not implement persisted snapshot listener anymore
 * I might found an **bug in the PassiveRole**: It seems the properties for a pending snapshot are not reset if persisting snapshot or calling the listeners fails with an exception
 * On the replay state machine we need to make sure that concurrent replays are not allowed. This can happen due to the commit listener. We need a flag/marker to avoid that. I had an bug where one event was applied to the state but the state changes was not yet committed, the next apply was already scheduled which caused an endless loop in NPE exceptions.

### Benchmark

I run another benchmark after my latest fixes and it looks now way better. It seems we reach almost the same throughput. I haven't tested yet fail-over. Other metrics like snapshoting and resource consumption look also good.


![fixed-general](https://user-images.githubusercontent.com/2758593/123280704-4ab91e80-d509-11eb-9044-9b320770355f.png)
![fixed-res](https://user-images.githubusercontent.com/2758593/123280707-4ab91e80-d509-11eb-890d-2275e11f30d3.png)
![fixed-compact](https://user-images.githubusercontent.com/2758593/123280700-4a208800-d509-11eb-8c0e-bd2e72c5f703.png)
![fixed-snapshot](https://user-images.githubusercontent.com/2758593/123280708-4b51b500-d509-11eb-9a9a-228fc3756249.png)




### Next Steps

 * Think about exporters. Either sending the state or writing an event.
 * Run some more benchmarks

**Comment by Zelldon** [comment]
### Day Five 25.05.2021

Today I was able to implement the last part, the exporter position distribution. I created a new SBE message, which is send in the ExporterDirector regulary (every 15 s). The follwer installs a service, which consumes the exporter positions messages and stores them in the ExportState. This makes it possible to distribute the positions from the Leader to the followers and the followers can respect these positions on compaction.

**Example Log:**
```
10:57:55.532 [] [Broker-0-Exporter-Satellite-1] DEBUG io.camunda.zeebe.broker.exporter - [exporterState-1] Received new exporter state {data-delete-test-exporter=292}
10:57:55.548 [] [Broker-1-Exporter-Satellite-1] DEBUG io.camunda.zeebe.broker.exporter - [exporterState-1] Received new exporter state {data-delete-test-exporter=292}
10:57:55.552 [Broker-2-SnapshotDirector-1] [Broker-2-zb-fs-workers-1] INFO  io.camunda.zeebe.logstreams.snapshot - Finished taking temporary snapshot, need to wait until last written event position 292 is committed, current commit position is 292. After that snapshot will be committed.
10:57:55.559 [Broker-1-SnapshotDirector-1] [Broker-1-zb-fs-workers-0] INFO  io.camunda.zeebe.logstreams.snapshot - Finished taking temporary snapshot, need to wait until last written event position 291 is committed, current commit position is 291. After that snapshot will be committed.
10:57:55.582 [Broker-0-SnapshotDirector-1] [Broker-0-zb-fs-workers-1] INFO  io.camunda.zeebe.logstreams.snapshot - Finished taking temporary snapshot, need to wait until last written event position 291 is committed, current commit position is 291. After that snapshot will be committed.
```

Furthermore I run today some benchmarks were I was able to get some more insights and to fix some smaller bugs. 
I will add later a summary of the POC.

### Findings 

 * In the PassiveRole, we added on the install request the transition to follower, to reset our services. This is problematic if the pending snapshot properties are not reseted before, since the snapshot might get deleted (since it will aborted on stop).

**Comment by Zelldon** [comment]
# POC Summary

In this [POC](https://github.com/camunda-cloud/zeebe/tree/zell-7328-poc-state-on-followers) we were able to show that it is possible to implement the concept from the [ZEP Draft](https://github.com/zeebe-io/enhancements/pull/16) without having significant impact on the processing throughput. 

We were able to build state on followers and remove the need of regularly replicating snapshots by the leader. It allows us to switch fast from a follower to a Leader role, and start processing, which should reduce the process execution latency during fail over.

These changes can reduce the maintenance cost as well, since we are able to reuse code, which we already use for the Leader and don't need to maintain anymore the push snapshot replication.

## Benchmark

I run at the end of the POC a benchmark in order to verify whether the system works as expected and
see what kind of impact the changes have. Especially throughput, latency and fail over latency was
interesting to see.

![poc-general](https://user-images.githubusercontent.com/2758593/123431353-726fbb80-d5c9-11eb-91ef-a58d1262f163.png)

We can see in the metrics that we are able to reach the same throughput as in our base benchmarks (200 PI/s). We haven't gone to the limit, which means we might see an impact on higher load but with our current benchmark we see no difference. The latency looks similar to our base as well.

![poc-latency](https://user-images.githubusercontent.com/2758593/123431411-83203180-d5c9-11eb-9f51-66eda3ad78ce.png)

The resource consumption looks also ok, much better than some days before.

![poc-res](https://user-images.githubusercontent.com/2758593/123431433-89aea900-d5c9-11eb-8ab1-8930c9114a72.png)

We can see that on all nodes and partitions snapshots are taken.

![poc-snapshot](https://user-images.githubusercontent.com/2758593/123431452-8fa48a00-d5c9-11eb-8c53-ef4ae81d9539.png)

The compaction is also working as expected.

![poc-atomix](https://user-images.githubusercontent.com/2758593/123431469-93d0a780-d5c9-11eb-8703-1c7f02ba46b8.png)

Interesting for the POC was the fail over time. With the fail over time, we mean the time from stopping processing on the old leader of partition X and starting processing on the new leader of partition X. We expected that we can reduce the fail over time. We were not able to show that directly in the metrics.

![restart-general](https://user-images.githubusercontent.com/2758593/123431520-a0550000-d5c9-11eb-943f-6dc73757bfba.png)

If we take a closer look, the problem seems to lie in RAFT itself.

![restart-heartbeats](https://user-images.githubusercontent.com/2758593/123431541-a6e37780-d5c9-11eb-87d2-2e358fc28116.png)


After Taking a look at the logs we can see that the nodes miss lot of appends, which causes followers to lag behind. This means on taking a leader down, the cluster can't elect a new leader until the old leader is at least restarted. We saw that polls not succeed for longer then 2 minutes.

**BUT** [If one node become Candidate (and Leader later) it was able to starting processing in ~ 1
second.](https://console.cloud.google.com/logs/viewer?interval=JUMP_TO_TIME&authuser=0&project=zeebe-io&minLogLevel=0&expandAll=false&timestamp=2021-06-25T11:33:19.037398000Z&customFacets=&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22zell-state-on-followers%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fcomponent%3D%22broker%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22zell-state-on-followers%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fmanaged-by%3D%22Helm%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22zeebe-cluster-helm%22%0Alabels.%22k8s-pod%2Fstatefulset_kubernetes_io%2Fpod-name%22%3D%22zell-state-on-followers-zeebe-0%22&scrollTimestamp=2021-06-25T11:33:22.730570000Z&dateRangeStart=2021-06-25T10:33:00.573Z&dateRangeEnd=2021-06-25T12:33:00.573Z&pinnedLogId=rpabr4mpopn2q9p6&pinnedLogTimestamp=2021-06-25T11:33:22.520958Z)

```
I 2021-06-25T11:33:22.520958Z RaftServer{raft-partition-partition-3} - Transitioning to CANDIDATE 
I 2021-06-25T11:33:22.522084Z RaftServer{raft-partition-partition-3}{role=CANDIDATE} - Starting election 
D 2021-06-25T11:33:22.530500Z Partition role transitioning from FOLLOWER to CANDIDATE in term 3 
I 2021-06-25T11:33:22.545709Z RaftServer{raft-partition-partition-3} - Transitioning to LEADER 
I 2021-06-25T11:33:22.546610Z RaftServer{raft-partition-partition-3} - Found leader 0 
D 2021-06-25T11:33:22.561829Z Partition role transitioning from CANDIDATE to LEADER in term 3 
D 2021-06-25T11:33:22.562195Z Closing Zeebe-Partition-3: ExporterSatelliteService 
D 2021-06-25T11:33:22.564373Z Closing Zeebe-Partition-3: ExporterSatelliteService closed successfully 
D 2021-06-25T11:33:22.564624Z Closing Zeebe-Partition-3: RocksDB metric timer 
D 2021-06-25T11:33:22.564893Z Closing Zeebe-Partition-3: RocksDB metric timer closed successfully 
D 2021-06-25T11:33:22.565083Z Closing Zeebe-Partition-3: AsyncSnapshotDirector 
D 2021-06-25T11:33:22.566324Z Closing Zeebe-Partition-3: AsyncSnapshotDirector closed successfully 
D 2021-06-25T11:33:22.566542Z Closing Zeebe-Partition-3: StreamProcessor 
D 2021-06-25T11:33:22.567627Z Closing Zeebe-Partition-3: StreamProcessor closed successfully 
D 2021-06-25T11:33:22.567630Z Closed stream processor controller Broker-0-StreamProcessor-3. 
D 2021-06-25T11:33:22.567859Z Closing Zeebe-Partition-3: ZeebeDb 
D 2021-06-25T11:33:22.568008Z Closing Zeebe-Partition-3: ZeebeDb closed successfully 
D 2021-06-25T11:33:22.568122Z Closing Zeebe-Partition-3: logstream 
I 2021-06-25T11:33:22.568266Z Close appender for log stream logstream-raft-partition-partition-3 
D 2021-06-25T11:33:22.569062Z Dispatcher closed 
I 2021-06-25T11:33:22.569407Z On closing logstream logstream-raft-partition-partition-3 close 1 readers 
D 2021-06-25T11:33:22.569580Z Closing Zeebe-Partition-3: logstream closed successfully 
D 2021-06-25T11:33:22.569697Z Partition 3 closed all previous open resources, before transitioning. 
D 2021-06-25T11:33:22.638315Z Configured log appender back pressure at partition 3 as AppenderVegasCfg{initialLimit=1024, maxConcurrency=32768, alphaLimit=0.7, betaLimit=0.95}. Window limiting is disabled 
D 2021-06-25T11:33:22.639197Z Recovering state of partition 3 from snapshot 
I 2021-06-25T11:33:22.639924Z Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@ee31df8)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@7de99f45, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@10d98228, configuration: Configuration(false)] 
I 2021-06-25T11:33:22.642688Z Recovered state of partition 3 from snapshot at position 6772750 
I 2021-06-25T11:33:22.643529Z Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@6f0774e7)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@791db822, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@483f03f3, configuration: Configuration(false)] 
I 2021-06-25T11:33:22.645204Z Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@142b101e)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@405ffe30, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@549792b1, configuration: Configuration(false)] 
I 2021-06-25T11:33:22.646834Z Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@6487334b)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@28606425, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@17f1a167, configuration: Configuration(false)] 
I 2021-06-25T11:33:22.649502Z Replay starts in mode UNTIL_END 
D 2021-06-25T11:33:22.650595Z Recovering exporter from snapshot 
D 2021-06-25T11:33:22.650609Z Partition 3 transition complete, installed 5 resources! 
I 2021-06-25T11:33:22.653365Z Processor finished replay. 
```

This result shows that in general it worked as we wanted and expected (the fast switch between the Follower and Leader partition services), but we need to find out why the followers lagging so much behind. It is quite likely that this might be due to some changes we did in the POC.

To summarize we can say the POC was a success, we were able to find answers to the following blocking questions:

1. How to rebuild the blacklist?
   See https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-865052187
1. How to implement continuous replay on followers? Do we get commit updates?
   See https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-867132086
1. What to do when the follower got an installation request?
   See https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-867683162
1. How we can distribute the exported positions to the
   followers? https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-868464316

Based on the knowledge we got, I think we can make a good plan for implementing the solution and
complete the reference section of the ZEP https://github.com/zeebe-io/enhancements/pull/16.

## Guide to Implement

In my opinion, to not overlap (conflict) too much with the bootstrap rewrite topic, we should try to complete that first. Ideally we synchronize to make sure the bootstrap logic, fulfills our needs (switching between Leader and Follower efficient).

In order to still make progress we could start to prepare for building state on followers and wire the things at the end
together.

In the following I will list what we need to do:

1. Write a ReplayStateMachine, which does replay until the end of the log.
1. Implement replay of ErrorEvents
    * blacklist state needs to be adjusted
1. Replace the ReprocessingStateMachine
1. Rewrite commit listeners in LogStream, should use the Raft commit listeners
    * to allow commit listeners on followers as well
1. Support continuously Replay in the ReplayStateMachine
      1. via commit listeners
1. Adjust StreamProcessor such it starts either in replay or processing (with replay in the
   beginning) mode
      1. e.g. enum flag
1. Prepare ExporterPositions SBE message
1. Send positions regularly from ExporterDirector
1. Implement new ExporterService which consumes SBE messages and updates Exporter positions in the
   state
1. Implement follower transition on install request (in PassiveRole)
    1. allows to reinit our follower services 
1. (After Bootstrap topic is done) Reorganize bootstrapping of a partition, several things should
   always exist (base) some services/components need to be reinstalled on transition (on role change) (see ZEP or POC)
    1. Implement the five transitions BoostrapLeader, BootstrapFollower, Leader-To-Follower, Follower-To-Leader and Follower-To-Follower.
    1. Make sure that the snapshots are not broadcast anymore

## Next Steps

* Review POC, with the team
* Enhance ZEP, add reference section
* Review ZEP
* If accepted implement proposal

**Comment by pihme** [comment]
What is the Follower-to-Follower transition?

**Comment by Zelldon** [comment]
See https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-867683162 InstallRequest handling

**Comment by Zelldon** [comment]
The POC is done, see [summary above](https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-868464316). I updated the ZEP https://github.com/zeebe-io/enhancements/pull/16. After the ZEP is accepted I will do a break down (similar I did in the summary) and create separate issues for it.
